fish-rust/src/job_group.rs

```diff
@@ -0,0 +1,352 @@
+    pub fn get_cancel_signal(&self) -> Option<NonZeroI32> {
+        match self.signal.load(Ordering::Relaxed) {
+            0 => None,
+            s => Some(NonZeroI32::new(s).unwrap()),
+        }
+    }
+
+    /// Gets the cancellation signal or `0` if none.
+    pub fn get_cancel_signal_ffi(&self) -> i32 {
+        // Legacy C++ code expects a zero in case of no signal.
+        self.get_cancel_signal().map(|s| s.into()).unwrap_or(0)
+    }
+
+    /// Mark that a process in this group got a signal and should cancel.
+    pub fn cancel_with_signal(&self, signal: NonZeroI32) {
+        // We only assign the signal if one hasn't yet been assigned. This means the first signal to
+        // register wins over any that come later.
+        self.signal
+            .compare_exchange(0, signal.into(), Ordering::Relaxed, Ordering::Relaxed)
+            .ok();
+    }
+
+    /// Mark that a process in this group got a signal and should cancel
+    pub fn cancel_with_signal_ffi(&self, signal: i32) {
+        self.cancel_with_signal(signal.try_into().expect("Invalid zero signal!"));
+    }
+
+    /// Set the pgid for this job group, latching it to this value. This should only be called if
+    /// job control is active for this group. The pgid should not already have been set, and should
+    /// be different from fish's pgid. Of course this does not keep the pgid alive by itself.
+    ///
+    /// Note we need not be concerned about thread safety. job_groups are intended to be shared
+    /// across threads, but any pgid should always have been set beforehand, since it's set
+    /// immediately after the first process launches.
+    ///
+    /// As such, this method takes `&mut self` rather than `&self` to enforce that this operation is
+    /// only available during initial construction/initialization.
+    pub fn set_pgid(&mut self, pgid: libc::pid_t) {
+        assert!(
+            self.wants_job_control(),
+            "Should not set a pgid for a group that doesn't want job control!"
+        );
+        assert!(pgid >= 0, "Invalid pgid!");
+        assert!(self.pgid.is_none(), "JobGroup::pgid already set!");
+
+        self.pgid = Some(pgid);
+    }
+
+    /// Returns the value of [`JobGroup::pgid`]. This is never fish's own pgid!
+    pub fn get_pgid(&self) -> Option<libc::pid_t> {
+        self.pgid
+    }
+
+    /// Returns the value of [`JobGroup::pgid`] in a `UniquePtr<T>` to take the place of an
+    /// `Option<T>` for ffi purposes. A null `UniquePtr` is equivalent to `None`.
+    pub fn get_pgid_ffi(&self) -> cxx::UniquePtr<pgid_t> {
+        match self.pgid {
+            Some(value) => UniquePtr::new(pgid_t { value }),
+            None => UniquePtr::null(),
+        }
+    }
+
+    /// Returns the current terminal modes associated with the `JobGroup` for ffi purposes.
+    unsafe fn get_modes_ffi(&self, size: usize) -> *const u8 {
+        assert_eq!(
+            size,
+            core::mem::size_of::<libc::termios>(),
```

Comment:
nit: we should be consistent about whether we use`std::mem` or `core::mem`


---

I haven't reviewed in detail but this seems nice :+1: